### PR TITLE
[Feature] Support Intern-S2-Preview training on NPU

### PR DESCRIPTION
**Pull request by CyCle1024**
## Summary

This PR adds NPU training support for Intern-S2-Preview while keeping the shared model and CUDA operator contracts stable.

- Improve MTP recomputation, FSDP scheduling, parameter residency, and LM-head prefetch control.
- Preserve device Tensor sequence boundaries and pass NPU CPU-list metadata through explicit parameters.
- Add Triton-Ascend fused cross-entropy for large-vocabulary training.
- Skip tail-padding tokens in the standard MoE dispatch path and cache the required indices.
- Add self-contained NPU GatedDeltaNet causal-convolution, gated-delta-rule, RMSNormGated, metadata, and sequence-parallel support.
- Integrate NPU CPU-affinity binding into the trainer lifecycle and retain host IRQ configuration helpers.
- Remove the dependency on a version-specific private DTensor shard API.

## Design notes

- Public cumulative-length arguments remain device tensors. NPU-only CPU lists and chunk metadata use separate, explicit parameters, so CUDA kernels do not receive Python lists.
- GatedDeltaNet uses `[B, T, H, D]` as its public layout. The NPU wrappers preserve native head-first storage with transpose views where possible and materialize only when required by a fixed-layout kernel.
- Normal GatedDeltaNet forward keeps one mixed-QKV convolution. Sequence-parallel forward gathers the global packed-token axis, shards heads, and runs separate Q/K/V convolutions, matching the NPU kernel geometry.
- GatedDeltaNet metadata is prepared once per sequence context and reused across layers. Cache reuse requires identical cumulative sequence boundaries, device, token count, and kernel block configuration.
- Backend-specific implementations and dispatch stay under `xtuner/v1/ops`; model modules keep device-independent logical interfaces.
- CUDA keeps its existing tensor inputs, loss-reduction behavior, and time-major GatedDeltaNet contract.

## Known limitations

- `skip_dispatch_pad_tokens` currently covers the standard single-context MoE forward, not the intra-layer multi-microbatch/domino-EP path.
- GatedDeltaNet currently requires `B == 1`, `num_value_heads` divisible by `num_key_heads`, and equal key/value head dimensions.
- In normal non-SP NPU GatedDeltaNet backward, PyTorch's generic mixed-QKV split backward produces a time-major contiguous gradient. The fixed-stride causal-convolution backward kernel therefore needs one additional layout materialization compared with the native-layout Tina prototype. The SP path does not have this mixed-QKV split. A stride-preserving custom autograd split is the intended follow-up optimization.
- Dynamic device-task and IRQ discovery is used by `export_bind_config()`. Trainer `run()` uses the packaged binding configuration; host IRQ setup remains an explicit launch-script operation.

### CODE REVIEW

File: xtuner/v1/train/trainer.py [hunk 1]
```diff
@@ -2062,7 +2062,10 @@ def _resolve_load_checkpoint_cfg(
 
     def _load_checkpoint(self):
         load_checkpoint_cfg: LoadCheckpointConfig = self._load_checkpoint_cfg
+        if DEVICE == "npu":
+            from xtuner.v1.utils import npu_cpu_binder
```
Comment (CyCle1024) [inline xtuner/v1/train/trainer.py]: move to another place

File: xtuner/v1/optim/muon.py [hunk 2]
```diff
@@ -1595,6 +1605,6 @@ def zeropower_via_newtonschulz5(G: Tensor, epsilon: float = 1e-7, num_experts: i
         X = X.mT
 
     # Reshape back to original shape: (num_experts, M, N) -> (num_experts * M, N)
-    X = X.view(original_shape)
+    X = X.reshape(num_experts * X.size(-2), X.size(-1))
```
Comment (CyCle1024) [inline xtuner/v1/optim/muon.py]: Rollback

File: xtuner/v1/ops/rms_norm/__init__.py [hunk 3]
```diff
@@ -63,11 +70,14 @@ def get_zero_centered_rms_norm_fn() -> RMSNormProtocol:
         else:
             return native_zero_centered_rms_norm
     elif device == "npu":
+        if os.getenv("XTUNER_USE_NATIVE_RMSNORM", "1") == "0":
+            raise NotImplementedError("Zero-centered RMSNorm is not implemented in triton")
+        else:
+            return zero_centered_rms_norm_npu
+        # def _not_implemented(*args, **kwargs):
```
Comment (CyCle1024) [inline xtuner/v1/ops/rms_norm/__init__.py]: remove comment

File: xtuner/v1/data_proto/sequence_context.py [hunk 4]
```diff
@@ -217,29 +241,36 @@ def from_input_ids(
         for ids in input_ids:
             assert ids.shape[0] == 1, "input_ids must have batch size of 1"
         num_tokens = [x.numel() for x in input_ids]
+        cu_seq_lens_list = [0, *itertools.accumulate(num_tokens)]
 
-        cu_seq_lens = cast(torch.IntTensor, torch.cumsum(torch.LongTensor([0] + num_tokens), dim=0).to(device).int())
+        cu_seq_lens = cast(torch.IntTensor, torch.tensor(cu_seq_lens_list, dtype=torch.int32, device=device))
```
Comment (CyCle1024) [inline xtuner/v1/data_proto/sequence_context.py]: use torch cumsum

File: xtuner/v1/data_proto/sequence_context.py [hunk 5]
```diff
@@ -249,14 +280,24 @@ def split(self, sequence_parallel_mesh: DeviceMesh | None = None) -> Self:
...
+                    new_cu_seq_lens_list = [
+                        *self.cu_seq_lens_q_list,
+                        self.cu_seq_lens_q_list[-1] + new_padding,
+                    ]
             else:
                 new_cu_seq_lens = self.cu_seq_lens_q.clone()
+                new_cu_seq_lens_list = self.cu_seq_lens_q_list.copy()
             new_cu_seq_lens = cast(torch.IntTensor, new_cu_seq_lens)
-            new_max_length = cast(int, max(self.seq_lens_q.max().item(), new_padding))
+            max_seq_length = max(
```
Comment (CyCle1024) [inline xtuner/v1/data_proto/sequence_context.py]: Rollback

File: xtuner/v1/model/moe/moe.py [hunk 6]
```diff
@@ -833,9 +836,9 @@ def _forward(
         self._mark_dynamic(seq_ctx)
         balancing_ctx, z_ctx = self._extract_aux_loss_ctx(loss_ctx)
         # Hoisted out of the per-layer accumulate path: mask is constant across layers.
-        nonpad_indices = torch.nonzero(seq_ctx.mask, as_tuple=True)[1]
+        nonpad_indices = seq_ctx.nonpad_indices
         non_pad_token = nonpad_indices.numel()
-        num_tokens_global, z_world_size = self._z_loss_dist_token_count(z_ctx, non_pad_token, seq_ctx.mask.device)
+        num_tokens_global, z_world_size = self._z_loss_dist_token_count(z_ctx, non_pad_token, nonpad_indices.device)
```
Comment (CyCle1024) [inline xtuner/v1/model/moe/moe.py:841]: Rollback

File: xtuner/v1/model/moe/moe.py [hunk 7]
```diff
@@ -845,7 +848,9 @@ def _forward(
                     seq_ctx=seq_ctx,
                 )
             else:
-                if int(os.getenv("XTUNER_ACTIVATION_OFFLOAD", "0")) == 1:
+                if int(os.getenv("XTUNER_ACTIVATION_OFFLOAD", "0")) == 1 and (
```
Comment (CyCle1024) [inline xtuner/v1/model/moe/moe.py]: Rollback

File: xtuner/v1/model/moe/moe.py [hunk 8]
```diff
@@ -885,14 +890,6 @@ def _forward(
                 output["hidden_states"].append(hidden_states)
 
         layer_hidden_states = hidden_states
-        hidden_states = self.norm(hidden_states)
-
-        # Get LM loss context from dict
```
Comment (CyCle1024) [inline xtuner/v1/model/moe/moe.py:890]: rollback

File: xtuner/v1/model/moe/moe.py [hunk 9]
```diff
@@ -909,10 +906,10 @@ def _forward(
                 ),
             )
             # MTP uses its own mask; main mask's non-pad indices do not apply.
-            mtp_nonpad_indices = torch.nonzero(mtp_seq_ctx.mask, as_tuple=True)[1]
+            mtp_nonpad_indices = mtp_seq_ctx.nonpad_indices
             mtp_non_pad_token = mtp_nonpad_indices.numel()
             mtp_num_tokens_global, mtp_z_world_size = self._z_loss_dist_token_count(
-                z_ctx, mtp_non_pad_token, mtp_seq_ctx.mask.device
+                z_ctx, mtp_non_pad_token, mtp_nonpad_indices.device
```
Comment (CyCle1024) [inline xtuner/v1/model/moe/moe.py:910]: rollback

File: xtuner/v1/model/moe/moe.py [hunk 10]
```diff
@@ -1287,24 +1297,42 @@ def fully_shard(
...
+            last_decoder_layer = list(self.layers.values())[-1]
+            last_decoder_layer.set_modules_to_forward_prefetch([first_mtp_layer])  # type: ignore
+
+            # Shared-weight MTP reuses one physical FSDP layer at every logical depth. A static
+            # forward-prefetch hook on that layer would therefore materialize the LM head after
+            # the first depth and keep it resident for all remaining depths. We intentionally
+            # leave the first post-MTP LM-head call to unshard on demand for now. A future
+            # optimization can explicitly prefetch the LM head in MTPBlock only after its final
+            # logical depth finishes.
+            if not mtp_config.share_weights:
```
Comment (CyCle1024) [inline xtuner/v1/model/moe/moe.py]: rollback

File: xtuner/v1/model/moe/moe.py [hunk 11]
```diff
@@ -1287,24 +1297,42 @@ def fully_shard(
...
-
-            if self.config.mtp_config is not None and self.config.mtp_config.num_layers > 0:
+            if mtp_config.num_layers > 0:
                 for prev_mtp_layer, next_mtp_layer in zip(
                     list(self.mtp_block.layers)[:-1],
                     list(self.mtp_block.layers)[1:],
                 ):
                     prev_mtp_layer.set_modules_to_forward_prefetch([next_mtp_layer])  # type: ignore
 
+            first_mtp_layer = self.mtp_block.layers[0]
```
Comment (CyCle1024) [inline xtuner/v1/model/moe/moe.py:1316]: last normal layer -> lmhead
lmhead -> mtp_first
and so on

File: xtuner/v1/model/moe/qwen3vl_text.py [hunk 12]
```diff
@@ -119,9 +119,9 @@ def _forward(
         self._mark_dynamic(seq_ctx)
         balancing_ctx, z_ctx = self._extract_aux_loss_ctx(loss_ctx)
         # Hoisted out of the per-layer accumulate path: mask is constant across layers.
-        nonpad_indices = torch.nonzero(seq_ctx.mask, as_tuple=True)[1]
+        nonpad_indices = seq_ctx.nonpad_indices
         non_pad_token = nonpad_indices.numel()
-        num_tokens_global, z_world_size = self._z_loss_dist_token_count(z_ctx, non_pad_token, seq_ctx.mask.device)
+        num_tokens_global, z_world_size = self._z_loss_dist_token_count(z_ctx, non_pad_token, nonpad_indices.device)
```
Comment (CyCle1024) [inline xtuner/v1/model/moe/qwen3vl_text.py]: rollback

File: xtuner/v1/module/dispatcher/torch_all2all.py [hunk 13]
```diff
@@ -526,10 +526,11 @@ def combine_preprocess(
         async_op: bool = False,
         decoding: bool = False,
     ) -> TorchAll2AllPreCombineResult:
-        hidden_states = unpermute(
-            hidden_states,
-            post_dispatched["row_ids_map"],
-        )
+        if len(hidden_states) != 0:
```
Comment (CyCle1024) [inline xtuner/v1/module/dispatcher/torch_all2all.py]:
```suggestion
        if hidden_states.numel() != 0:
```

**Review by CyCle1024 — COMMENTED** [review]
fix

**Review by jayhenry — APPROVED** [review]
